Reduce log spam on gossip and raft transport

spencerkimball opened the pull request:
This is necessary to avoid lots of spurious RPC errors on node startup
caused by the connection believing itself to be unhealthy and failing
everything fast. With thousands of ranges, this was quite confusing.

Remove spammy logging via use of circuit breakers to measure state
changes. In particular with gossip and raft transport.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8529)

<!-- Reviewable:end -->

spencerkimball committed:
Reduce log spam on gossip and raft transport

By checking whether a circuit breaker is tripped in raft transport,
we now fail fast in `SendAsync`, causing caller to drop the raft message
and report the replica unreachable to the raft group.

The asynchronous message queue processing employs `Breaker.Call` to
trip the breaker in the event that the node ID can't be resolved, the
GRPC connection can't be established, or the GRPC connection isn't
healthy. The circuit breaker is used to only log state transitions
for connections to nodes to avoid spammy logging.

Only log on raft transport state changes to unhealthy connections.

Also augmented the `grpcutil.IsCloseConnection` method to consider the
grpc error code of `Unavailable` as the connection being closed, which
removes more spammy logging.

Use a circuit breaker on gossip clients to only log state changes on
gossip clients to a single address.

tamird commented:
If you want GRPC dialing to block, why not use `grpc.WithBlock()`?

I've reviewed this only partially pending an answer to that question.

---

Reviewed 5 of 16 files at r1.
Review status: 5 of 16 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[kv/transport.go, line 123 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KP9C4ZioDB1sXpnRm2S:-KP9C4ZioDB1sXpnRm2T:btoxfsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/00354c9969c8e2a30163ef43255a5b9d29b67235/kv/transport.go#L123)):*
> ```Go
> 		conn, err := rpcContext.GRPCDial(replica.NodeDesc.Address.String())
> 		if err != nil {
> 			if errors.Cause(err) == circuit.ErrBreakerOpen {
> ```

why'd this change?

---

*[rpc/context_test.go, line 118 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KP9CGxibSlvM6QIW6YX:-KP9CGxibSlvM6QIW6YY:bbei766) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf4588f854fee6c2b7868ca7220fe00116acdcfa/rpc/context_test.go#L118)):*
> ```Go
> 	clientCtx.HeartbeatTimeout = 1 * time.Millisecond
> 	// Expect a failure on the initial dial as no heartbeat is available.
> 	if _, err := clientCtx.GRPCDial(remoteAddr); !testutils.IsError(err, "context deadline exceeded") {
> ```

seems like this can check the error directly `err == context.DeadlineExceeded`. ditto below

---

*[rpc/context_test.go, line 249 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KP9COPmlypJ9VSIRYzK:-KP9COPmlypJ9VSIRYzL:b-tqzpzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf4588f854fee6c2b7868ca7220fe00116acdcfa/rpc/context_test.go#L249)):*
> ```Go
> 
> 	clientCtx := newNodeTestContext(clock, stopper)
> 	// Make the intervals and timeouts shorter to speed up the tests.
> ```

this was intentionally not messing with timeouts to ensure that the failures here were observed due to transport failures, not timeouts. why did you change it?

---

*[rpc/context_test.go, line 405 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KP9CYbvZXJQOIJzIcqE:-KP9CYbvZXJQOIJzIcqF:b-4p2kkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf4588f854fee6c2b7868ca7220fe00116acdcfa/rpc/context_test.go#L405)):*
> ```Go
> 	// clock reading delay, not the timeout.
> 	clientCtx.HeartbeatTimeout = 20 * clientCtx.HeartbeatInterval
> 	stopper.RunWorker(func() {
> ```

what's this about? it is not safe to call `t.Fatal` in the non-main goroutine.

---

*[util/grpcutil/grpc_util.go, line 37 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KP9Bov56oudt1ut2NnK:-KP9Bov6R5-nOfAPd7sl:bbxf271) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf4588f854fee6c2b7868ca7220fe00116acdcfa/util/grpcutil/grpc_util.go#L37)):*
> ```Go
> 		grpc.ErrorDesc(err) == grpc.ErrClientConnClosing.Error() ||
> 		strings.Contains(err.Error(), "is closing") ||
> 		strings.Contains(err.Error(), "node unavailable") {
> ```

this is probably better checked as `grpc.Code(err) == codes.Unavailable`. why did it become necessary?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-:-KP9CelNnkb1WlrMFyEe:b-cvx2d8)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
I didn't realize that was available. Switching to it does fix the larger problem I was seeing with initial raft transport failures. I've updated the PR.

---

Review status: 4 of 15 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[kv/transport.go, line 123 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KP9C4ZioDB1sXpnRm2S:-KPA59WmKtNxv-0At13J:bln7yyk) ([raw file](https://github.com/cockroachdb/cockroach/blob/00354c9969c8e2a30163ef43255a5b9d29b67235/kv/transport.go#L123)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why'd this change?
</details>
There's no longer a breaker employed by the GRPCDial call so checking for this error doesn't make any sense.

---

*[util/grpcutil/grpc_util.go, line 37 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KP9Bov56oudt1ut2NnK:-KPA5JsGNkAnLl_KFN6j:b-vmatk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf4588f854fee6c2b7868ca7220fe00116acdcfa/util/grpcutil/grpc_util.go#L37)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is probably better checked as `grpc.Code(err) == codes.Unavailable`. why did it become necessary?
</details>
will that work? I'm checking for the error we get when a node is being shutdown and is in drain mode. Does that create the right GRPC error code?

---

*[rpc/context_test.go, line 118 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KP9CGxibSlvM6QIW6YX:-KPA5ZVpkOqaa3iR0-14:b69kmyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf4588f854fee6c2b7868ca7220fe00116acdcfa/rpc/context_test.go#L118)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> seems like this can check the error directly `err == context.DeadlineExceeded`. ditto below
</details>
This code has been removed.

---

*[rpc/context_test.go, line 249 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KP9COPmlypJ9VSIRYzK:-KPA5aV39N-P-kewkgyf:b-ll2wdd) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf4588f854fee6c2b7868ca7220fe00116acdcfa/rpc/context_test.go#L249)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this was intentionally not messing with timeouts to ensure that the failures here were observed due to transport failures, not timeouts. why did you change it?
</details>
This change has been backed out.

---

*[rpc/context_test.go, line 405 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KP9CYbvZXJQOIJzIcqE:-KPA5dD4a5W2fwsN7Wn1:bwrznim) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf4588f854fee6c2b7868ca7220fe00116acdcfa/rpc/context_test.go#L405)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what's this about? it is not safe to call `t.Fatal` in the non-main goroutine.
</details>
Ditto. This is gone now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-:-KPA4vjsk859VYzj7Rbp:b-ndxd9d)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 8 of 16 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 68 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPAADyMy7CEDgEOaf64:-KPAADyMy7CEDgEOaf65:b-fs02rw) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/gossip/client.go#L68)):*
> ```Go
> 		clientMetrics:         makeMetrics(),
> 		nodeMetrics:           nodeMetrics,
> 		breaker:               breaker,
> ```

ISTM that this can be removed - only `start` uses `breaker`, perhaps it could be passed into that function instead of the constructor.

---

*[gossip/client.go, line 78 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA9gMDlfIVSzK1Dz6H:-KPA9gMDlfIVSzK1Dz6I:bi9crvc) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/gossip/client.go#L78)):*
> ```Go
> 	ctx := context.TODO()
> 	nodeID := g.is.NodeID
> 	if c.breaker == nil {
> ```

how can this ever happen?

---

*[gossip/client.go, line 85 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA9r2fVFq_4UPLr3P_:-KPA9r2fVFq_4UPLr3Pa:b-6baa7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/00354c9969c8e2a30163ef43255a5b9d29b67235/gossip/client.go#L85)):*
> ```Go
> 		// asynchronous from the caller's perspective, so the only effect of
> 		// `WithBlock` here is blocking shutdown - at the time of this writing,
> 		// that ends ups up making `kv` tests take twice as long.
> ```

have you verified the impact on kv tests?

---

*[gossip/gossip.go, line 162 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA8PGtq88nbgqrvzwh:-KPA8PGtq88nbgqrvzwi:bpsok2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/gossip/gossip.go#L162)):*
> ```Go
> 		syncutil.Mutex
> 		clients  []*client
> 		breakers map[string]*circuit.Breaker
> ```

are these stored separately from clients so that they can outlive individual clients? that needs a comment.

---

*[rpc/context.go, line 190 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA8B2PsbYpJSqXvwGl:-KPA8B2PsbYpJSqXvwGm:b78g0db) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/rpc/context.go#L190)):*
> ```Go
> 		}
> 	}
> ```

this file now only contains whitespace changes, can you back them out?

---

*[server/node.go, line 623 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA70r1luT8H-iXd5EV:-KPA70r2Tn7dNOKGopJg:bd7d7gd) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/server/node.go#L623)):*
> ```Go
> 			select {
> 			case <-ticker.C:
> 				_ = n.computePeriodicMetrics(tick)
> ```

why is error checking no longer needed?

---

*[server/node.go, line 636 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA7S6E0qA2d_Llkfj_:-KPA7S6E0qA2d_Llkfja:bk7yyvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/server/node.go#L636)):*
> ```Go
> 	return n.stores.VisitStores(func(store *storage.Store) error {
> 		if err := store.ComputeMetrics(tick); err != nil {
> 			log.Warningf(context.TODO(), "%s: unable to compute metrics: %s", store, err)
> ```

this change trades error swallowing in `computeReplicationStatus` for the same thing here. why is this better? why not pass the error through?

---

*[storage/raft_transport.go, line 209 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA6bqtsfDC2D5_6aMZ:-KPA6bqtsfDC2D5_6aM_:b375nf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/storage/raft_transport.go#L209)):*
> ```Go
> 		addr, err = t.resolver(nodeID)
> 		if err != nil {
> 			return err
> ```

`errors.Wrap(err, "failed to resolve")` will give slightly more specific information

---

*[storage/raft_transport.go, line 212 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA6kFjIT_vjbpI5LvU:-KPA6kFjIT_vjbpI5LvV:bc895bc) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/storage/raft_transport.go#L212)):*
> ```Go
> 		}
> 		conn, err = t.rpcContext.GRPCDial(addr.String(), grpc.WithBlock())
> 		return err
> ```

`errors.Wrap(err, "failed to dial")`

---

*[storage/store.go, line 988 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA6KAkdbFnt72Yc3wG:-KPA6KAkdbFnt72Yc3wH:b-e8heha) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/storage/store.go#L988)):*
> ```Go
> 
> 		// Run metrics computation up front to populate initial statistics.
> 		_ = s.ComputeMetrics(-1)
> ```

why is error checking no longer desired?

---

*[storage/store.go, line 2604 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA99JfXy946bUfqKqt:-KPA99JfXy946bUfqKqu:b-xnjgz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/storage/store.go#L2604)):*
> ```Go
> 
> 	// If we're using RocksDB, log the sstable overview.
> 	if rocksdb, ok := s.engine.(*engine.RocksDB); ok && tick%100 == 0 {
> ```

there looks to be only one caller passing a non-constant value of `tick` - why not make the parameter an explicit "logSSTable" parameter?

---

*[util/grpcutil/grpc_util.go, line 37 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KP9Bov56oudt1ut2NnK:-KPA61PJMLI9FKRVrI-9:b-7o5ag3) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf4588f854fee6c2b7868ca7220fe00116acdcfa/util/grpcutil/grpc_util.go#L37)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> will that work? I'm checking for the error we get when a node is being shutdown and is in drain mode. Does that create the right GRPC error code?
</details>
Ah, no it doesn't. What changed to make this check required?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 78 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA9gMDlfIVSzK1Dz6H:-KPAowlCGPmcdJ3pgV0K:b8ofnu4) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/gossip/client.go#L78)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how can this ever happen?
</details>
from unittests

---

*[gossip/client.go, line 85 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA9r2fVFq_4UPLr3P_:-KPAoytyw6VrNvdo8Al3:b-kyeq5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/00354c9969c8e2a30163ef43255a5b9d29b67235/gossip/client.go#L85)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> have you verified the impact on kv tests?
</details>
I didn't mean to remove the comment. Restored.

---

*[server/node.go, line 623 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA70r1luT8H-iXd5EV:-KPApfBJrpwmkdGbIAsn:b-bvwhm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/server/node.go#L623)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why is error checking no longer needed?
</details>
Restored.

---

*[server/node.go, line 636 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA7S6E0qA2d_Llkfj_:-KPApj0VXNotPe9o76v3:baf3agd) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/server/node.go#L636)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this change trades error swallowing in `computeReplicationStatus` for the same thing here. why is this better? why not pass the error through?
</details>
I want to log the store which failed, but more importantly want to go through the all the stores to compute metrics without erroring out midway through the visitor.

---

*[storage/store.go, line 988 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA6KAkdbFnt72Yc3wG:-KPAqQXOjQU2e6NJ2OtA:b-oyjdmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/storage/store.go#L988)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why is error checking no longer desired?
</details>
Added logging. Returning an error is incorrect as failing to compute initial metrics doesn't mean the store wasn't started.

---

*[storage/store.go, line 2604 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA99JfXy946bUfqKqt:-KPAsAIV4DdnttfOY-rW:bbl0mdg) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/storage/store.go#L2604)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> there looks to be only one caller passing a non-constant value of `tick` - why not make the parameter an explicit "logSSTable" parameter?
</details>
Because I don't believe this will be the only thing we decide to log with varying periodicities.

---

*[util/grpcutil/grpc_util.go, line 37 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KP9Bov56oudt1ut2NnK:-KPAsJFE-Jd8eNb-qRp8:b5fcypy) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf4588f854fee6c2b7868ca7220fe00116acdcfa/util/grpcutil/grpc_util.go#L37)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Ah, no it doesn't. What changed to make this check required?
</details>
When running a big cluster, if you take a node down with control-C, every range connected to it starts spewing failures.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529)*
<!-- Sent from Reviewable.io -->

tamird commented:
I don't see any of the latest round of changes. Did you push?

---

Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-:-KPB8MSY3tjka2nds0u1:bxs2lsn)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
PTAL

petermattis commented:
Review status: 1 of 15 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[storage/raft_transport.go, line 235 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPF3NtavcnVIOBLskax:-KPF3NtavcnVIOBLskay:b-lojts3) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe1a638355941056fdb47a6434a45127ce9de1/storage/raft_transport.go#L235)):*
> ```Go
> 			return errors.Wrapf(err, "failed to resolve node %d -> address", nodeID)
> 		}
> 		conn, err = t.rpcContext.GRPCDial(addr.String(), grpc.WithBlock())
> ```

Blocking the dial call will in turn block `RaftTransport.SendAsync` which is called from `Replica.handleRaftReady -> Replica.sendRaftMessage`. I think it will be bad to block up raft processing waiting for network connections to open. We use to open the node connection for raft processing inside of `processQueue`. Perhaps it is time to move back to that approach.

Also, I feel like we've gone back and forth on blocking during dialing (at least, I know I tried it at one point). Let's add a comment here explaining why this is necessary.

---

*[storage/store.go, line 2604 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA99JfXy946bUfqKqt:-KPF5N6Fld9DNkg34ECx:b-g9cw0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/storage/store.go#L2604)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Because I don't believe this will be the only thing we decide to log with varying periodicities.
</details>
Don't we always want to compute the read amplification metric, but only log it when `tick%100 == 0`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 1 of 16 files at r1, 15 of 15 files at r3.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 74 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPF5yQzaaFqMP8pnxUW:-KPF5yR-YC0mn4JDgFed:bs1tma5) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe1a638355941056fdb47a6434a45127ce9de1/gossip/client.go#L74)):*
> ```Go
> // processing in a goroutine and returns immediately.
> func (c *client) start(g *Gossip, disconnected chan *client, rpcCtx *rpc.Context, stopper *stop.Stopper, breaker *circuit.Breaker) {
> 	ctx, cancel := context.WithCancel(context.TODO())
> ```

Put this inside the worker callback now that we no longer need it for logging before the worker starts.

---

*[storage/raft_transport.go, line 235 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPF3NtavcnVIOBLskax:-KPF7GeAytBoa_gR6a8x:b-dm7x3) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe1a638355941056fdb47a6434a45127ce9de1/storage/raft_transport.go#L235)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Blocking the dial call will in turn block `RaftTransport.SendAsync` which is called from `Replica.handleRaftReady -> Replica.sendRaftMessage`. I think it will be bad to block up raft processing waiting for network connections to open. We use to open the node connection for raft processing inside of `processQueue`. Perhaps it is time to move back to that approach.
> 
> Also, I feel like we've gone back and forth on blocking during dialing (at least, I know I tried it at one point). Let's add a comment here explaining why this is necessary.
</details>
+1 to making the connection inside `processQueue` if we're going to block.

I'm not sure I understand why blocking is necessary here (and only here), though. In GRPC, the health of a connection is not binary. It starts in state "connecting", which causes fail-fast RPCs to block until it becomes either healthy or unhealthy, instead of failing immediately. You shouldn't see the connection as unhealthy until there has been a real failure. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 15 of 15 files at r3.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 78 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPA9gMDlfIVSzK1Dz6H:-KPFBa3ZSLm-n4iziUDO:b-wj0z61) ([raw file](https://github.com/cockroachdb/cockroach/blob/13fcf7de6453ef8e00f65f4c310979e1fb15a978/gossip/client.go#L78)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> from unittests
</details>
Can we fix the unit tests? I'd prefer to avoid this kind of thing since it makes this code harder to read.

I left a comment on each instance, there appear to be only 3.

---

*[gossip/client_test.go, line 138 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPFGqjUSFK44hS61gbz:-KPFGqjUSFK44hS61gc-:b1cra5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe1a638355941056fdb47a6434a45127ce9de1/gossip/client_test.go#L138)):*
> ```Go
> 		case client := <-disconnected:
> 			// If the client wasn't able to connect, restart it.
> 			client.start(gossip[client], disconnected, rpcContext, stopper, nil)
> ```

pass a breaker

---

*[gossip/client_test.go, line 268 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPFGri4nDfk1-UIq_Zz:-KPFGri4nDfk1-UIq__-:b1cra5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe1a638355941056fdb47a6434a45127ce9de1/gossip/client_test.go#L268)):*
> ```Go
> 		case <-disconnected:
> 			// The client hasn't been started or failed to start, loop and try again.
> 			c.start(local, disconnected, rpcContext, stopper, nil)
> ```

pass a breaker

---

*[gossip/gossip.go, line 162 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPFGhD5xYsNY0HJEvaq:-KPFGhD6J-h-Xa5b4v0W:b44w2fc) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe1a638355941056fdb47a6434a45127ce9de1/gossip/gossip.go#L162)):*
> ```Go
> 		syncutil.Mutex
> 		clients []*client
> 		// One breaker per client per execution.
> ```

"execution" is not really clear - you mean that this is for the life of the process, right?

---

*[gossip/gossip_test.go, line 197 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPFG7JlvUrnhLIId2Br:-KPFG7JlvUrnhLIId2Bs:b-x38qif) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe1a638355941056fdb47a6434a45127ce9de1/gossip/gossip_test.go#L197)):*
> ```Go
> 		for {
> 			c := newClient(&local.is.NodeAddr, makeMetrics())
> 			c.start(peer, disconnectedCh, peer.rpcContext, stopper, nil)
> ```

pass a real breaker to avoid the nil check in "real" code

---

*[rpc/context.go, line 111 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPFFwHHCRqVTHs9Wdof:-KPFFwHHCRqVTHs9Wdog:b-hqb2me) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe1a638355941056fdb47a6434a45127ce9de1/rpc/context.go#L111)):*
> ```Go
> 		clock: ctx.localClock,
> 	}
> 	ctx.dialCtx, ctx.dialCancel = context.WithCancel(context.Background())
> ```

this context is not really specific to dialing, so this name could be better. probably just `ctx` and `cancel`?

---

*[storage/raft_transport.go, line 239 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPFEDEqb4FbrytU9JxV:-KPFEDEqb4FbrytU9JxW:bfuo1rn) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe1a638355941056fdb47a6434a45127ce9de1/storage/raft_transport.go#L239)):*
> ```Go
> 	}, 0); err != nil {
> 		if consecFailures == 0 {
> 			log.Warningf(context.TODO(), "failed to connect node %s: %s", nodeID, err)
> ```

you're double logging the node ID now since it's already in `err`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 7 of 15 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[rpc/context.go, line 111 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPFFwHHCRqVTHs9Wdof:-KPIXDbqOgxyFTOPrtf5:bdau45o) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe1a638355941056fdb47a6434a45127ce9de1/rpc/context.go#L111)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this context is not really specific to dialing, so this name could be better. probably just `ctx` and `cancel`?
</details>
calling it `ctx` is really confusing. This struct is `rpc.Context` and it embeds a `base.Context`.

---

*[storage/raft_transport.go, line 235 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPF3NtavcnVIOBLskax:-KPIaA8tp_OQ6j1_W7wn:bh0gq26) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe1a638355941056fdb47a6434a45127ce9de1/storage/raft_transport.go#L235)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> +1 to making the connection inside `processQueue` if we're going to block.
> 
> I'm not sure I understand why blocking is necessary here (and only here), though. In GRPC, the health of a connection is not binary. It starts in state "connecting", which causes fail-fast RPCs to block until it becomes either healthy or unhealthy, instead of failing immediately. You shouldn't see the connection as unhealthy until there has been a real failure. 
</details>
Ugh I agree about blocking `Replica.handleRaftReady`. That's no good. The point of getting to this failure before `processQueue` is to return `false` from `SendAsync`, which would cause raft messages to be quickly and silently dropped when the transport is broken. Previously, we were just queuing them up. This would mean that every replica in the system talking to a down node would immediately queue up 100 raft messages.

I've changed things around here so we're back to using the non-blocking `Dial`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/raft_transport.go, line 365 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPJtW9KuPhrCweqND7z:-KPJtW9KuPhrCweqND8-:b-mktwe1) ([raw file](https://github.com/cockroachdb/cockroach/blob/da38070e41802bb9fc042723e3a018d215736ab9/storage/raft_transport.go#L365)):*
> ```Go
> 			t.rpcContext.Stopper.RunWorker(func() {
> 				if err := t.processQueue(ch, conn); err != nil && !grpcutil.IsClosedConnection(err) {
> 					log.Warningf(context.TODO(), "range=%s: raft transport stream to node %d: %s", req.RangeID, toNodeID, err)
> ```

there's no verb here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 8 of 8 files at r4, 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[storage/raft_transport.go, line 235 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPF3NtavcnVIOBLskax:-KPK8TrgmLw7D2aE8Y_f:b-yb9f1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe1a638355941056fdb47a6434a45127ce9de1/storage/raft_transport.go#L235)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Ugh I agree about blocking `Replica.handleRaftReady`. That's no good. The point of getting to this failure before `processQueue` is to return `false` from `SendAsync`, which would cause raft messages to be quickly and silently dropped when the transport is broken. Previously, we were just queuing them up. This would mean that every replica in the system talking to a down node would immediately queue up 100 raft messages.
> 
> I've changed things around here so we're back to using the non-blocking `Dial`.
</details>
If you're not blocking then you're probably not going to be able to get an error from `GRPCDial`, so you don't really gain anything by putting it inside the circuit breaker (it's normally just a cache hit in rpc.Context). 

---

*[storage/raft_transport.go, line 365 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPK5UO_0at_9xvMIv5y:-KPK5UO_0at_9xvMIv5z:b-oy2lz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec88b99a75b1dabc17a21a8f268ecd6ac593395c/storage/raft_transport.go#L365)):*
> ```Go
> 			t.rpcContext.Stopper.RunWorker(func() {
> 				if err := t.processQueue(ch, conn); err != nil && !grpcutil.IsClosedConnection(err) {
> 					log.Warningf(context.TODO(), "range=%s: raft transport stream to node %d failed: %s", req.RangeID, toNodeID, err)
> ```

While you're here remove the `RangeID` from this message - it's misleading since many ranges share one queue. 

This will still be pretty spammy in the event that the connection is unhealthy in a way that `grpcutil.IsClosedConnection` doesn't match. That's what's currently happening on beta, where we're getting a "context deadline exceeded" error printed for every replica. We could add that to `IsClosedConnection` but I think it may be better to limit log spamminess with a circuit breaker for processQueue instead of blacklisting messages by their content.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[storage/raft_transport.go, line 235 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPF3NtavcnVIOBLskax:-KPKC2ImXtvg634FzzFT:b1a2m3g) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe1a638355941056fdb47a6434a45127ce9de1/storage/raft_transport.go#L235)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> If you're not blocking then you're probably not going to be able to get an error from `GRPCDial`, so you don't really gain anything by putting it inside the circuit breaker (it's normally just a cache hit in rpc.Context). 
</details>
Well it returns an error in the event that the context it's using has already been canceled, which spammed the logs on shutdown.

---

*[storage/raft_transport.go, line 365 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPK5UO_0at_9xvMIv5y:-KPKCvahMJP6frIVI6OL:b-1ga7rd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec88b99a75b1dabc17a21a8f268ecd6ac593395c/storage/raft_transport.go#L365)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> While you're here remove the `RangeID` from this message - it's misleading since many ranges share one queue. 
> 
> This will still be pretty spammy in the event that the connection is unhealthy in a way that `grpcutil.IsClosedConnection` doesn't match. That's what's currently happening on beta, where we're getting a "context deadline exceeded" error printed for every replica. We could add that to `IsClosedConnection` but I think it may be better to limit log spamminess with a circuit breaker for processQueue instead of blacklisting messages by their content.
</details>
Removed `RangeID` from the log message.

Another circuit breaker isn't quite right here because the underlying grpc connection is still reconnecting in many cases that arise in practice. I've changed it to only log on a state change from a successful connection to a broken connection.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529)*
<!-- Sent from Reviewable.io -->

tamird commented:
I believe this is still waiting on https://github.com/grpc/grpc-go/pull/831. That said, I'm not sure the commit message still accurately describes what's being done in this PR. In fact, I'm not sure I still understand what's being done in this PR or what the motivation is. Could you update the commit?

bdarnell commented:
Review status: 14 of 15 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[storage/raft_transport.go, line 235 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPF3NtavcnVIOBLskax:-KPNvV57nVYjijOtwlAw:bmo9he7) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe1a638355941056fdb47a6434a45127ce9de1/storage/raft_transport.go#L235)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Well it returns an error in the event that the context it's using has already been canceled, which spammed the logs on shutdown.
</details>
OK, so that's one case, but in a production "down node" scenario you're getting an error from creating the stream in processQueue and not from getNodeConn. That means you're not returning false from SendAsync (and therefore not calling raftGroup.ReportUnreachable). This file needs a bigger refactoring than some tweaks to its log spamminess. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
The commit message in git was updated a while ago. Are you referring to the one in github?

---

Review status: 12 of 23 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[storage/raft_transport.go, line 235 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPF3NtavcnVIOBLskax:-KPQB_Q0eWbCeloI5KwM:bkfe7a6) ([raw file](https://github.com/cockroachdb/cockroach/blob/befe1a638355941056fdb47a6434a45127ce9de1/storage/raft_transport.go#L235)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> OK, so that's one case, but in a production "down node" scenario you're getting an error from creating the stream in processQueue and not from getNodeConn. That means you're not returning false from SendAsync (and therefore not calling raftGroup.ReportUnreachable). This file needs a bigger refactoring than some tweaks to its log spamminess. 
</details>
OK, I've addressed this point. The first one or several messages will get added to the queue while the connection is asynchronously resolved, dialed, and a stream is created. After the first failure, the breaker kicks in and then `false` is returned immediately from `SendAsync`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-:-KPO9yTddIjPvjVLkcFQ:b-8g62bo)*
<!-- Sent from Reviewable.io -->

spencerkimball commented on GLOCKFILE:
```diff
@@ -109,7 +109,7 @@ golang.org/x/sys a646d33e2ee3172a661fc09bca23bb4889a41bc8
 golang.org/x/tools 0e9f43fcb67267967af8c15d7dc54b373e341d20
 google.golang.org/appengine e951d3868b377b14f4e60efa3a301532ee3c1ebf
-google.golang.org/grpc c2c110d5cf950aef5e2af86b2f5d0a53400ff90b
+google.golang.org/grpc 8a81ddda27e5fdd3b56f3ce841b37dbf7387dd26
```
This picks up the `DialContext` method.

tamird commented:
There seem to be a bunch of unrelated queue changes here. Was that intended?

---

Reviewed 4 of 5 files at r5, 11 of 12 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[gossip/client.go, line 79 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQRvp_GLd_BV8LzTCT:-KPQRvp_GLd_BV8LzTCU:b-195x5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/gossip/client.go#L79)):*
> ```Go
> 	stopper *stop.Stopper,
> 	nodeID roachpb.NodeID,
> 	breaker *circuit.Breaker) {
> ```

please follow the style guide - the closing paren should be on its own line together with the opening brace:

```
  breaker *circuit.Breaker,
) {
```

---

*[rpc/context.go, line 89 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQSlgI7cyNAEliARgz:-KPQSlgI7cyNAEliARh-:b9plq6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/rpc/context.go#L89)):*
> ```Go
> 	HeartbeatTimeout  time.Duration
> 	HeartbeatCB       func()
> 	BreakerFactory    func() *circuit.Breaker
> ```

needs a comment, since there is already a NewBreaker method and it's not clear why both are needed (testing, I presume?)

---

*[rpc/context.go, line 192 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQSQnRjL9V4GuBcy6I:-KPQSQnSb2z5U31gj4Sj:bu0sobs) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/rpc/context.go#L192)):*
> ```Go
> 		}
> 		conn, err := grpc.DialContext(ctx.masterCtx, target, dialOpts...)
> 		ctx.conns.Lock()
> ```

why does this lock need to be held here? all concurrent callers to GRPCDial will block while this is executing.

---

*[rpc/context.go, line 211 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQSgJK565OYPbv7yKO:-KPQSgJK565OYPbv7yKP:b-ewjed1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/rpc/context.go#L211)):*
> ```Go
> 	})
> 
> 	ctx.conns.Lock()
> ```

this lock also doesn't appear necessary.

---

*[storage/client_raft_test.go, line 1175 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQSvq2PU9Mk4D9oAwN:-KPQSvq2PU9Mk4D9oAwO:b-r7bvyb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/storage/client_raft_test.go#L1175)):*
> ```Go
> 		}
> 
> 		// Verify old replica is GC'd.
> ```

does this have anything to do with the rest of this change?

---

*[storage/client_test.go, line 271 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQT95i4r9H_a7M2fCH:-KPQT95i4r9H_a7M2fCI:b-qih6wn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/storage/client_test.go#L271)):*
> ```Go
> 	if m.rpcContext == nil {
> 		m.rpcContext = rpc.NewContext(&base.Context{Insecure: true}, m.clock, m.transportStopper)
> 		// Create breaker options which retry very quickly and uses a "real"
> ```

s/uses/use/ ?

---

*[storage/client_test.go, line 273 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQT2egTHWi5YIkkc_a:-KPQT2egTHWi5YIkkc_b:b4qjwx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/storage/client_test.go#L273)):*
> ```Go
> 		// Create breaker options which retry very quickly and uses a "real"
> 		// clock so breakers retry without requiring manual clock, used as
> 		// part of test context, be incremented manually to untrip breaker.
> ```

s/be/to be/ ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-:-KPQTqGebGcyKb3aTAWy:b2kcou3)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 11 of 12 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[storage/client_raft_test.go, line 1170 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQTTWbt7Q09J8tkjfm:-KPQTTWbt7Q09J8tkjfn:b-t3nuab) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/storage/client_raft_test.go#L1170)):*
> ```Go
> 	mtc.restartStore(1)
> 	if removeBeforeTruncateAndReAdd {
> 		oldRng, err := mtc.stores[1].GetReplica(rangeID)
> ```

This looks racy - the replica could have been gc'd between the restartStore and this point. Elsewhere when we check for this we just call GetReplica in a SucceedsSoon and wait for it to return "not found". 

---

*[storage/queue.go, line 280 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQPAWDG2SE_BG2GzqB:-KPQPAWDG2SE_BG2GzqC:b5d9xxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/storage/queue.go#L280)):*
> ```Go
> // error if the replica was not added. Returns whether the replica was
> // added (true), or was already added (false), and an error if the
> // replica was not added.
> ```

This comment sounds like you would never see (false, nil), but that's not true. "Returns (true, nil) if the replica was added, (false, nil) if the replica was already present, and (false, err) if the replica could not be added for any other reason".

---

*[storage/raft_transport.go, line 232 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQVKw6e5OocA43HrZS:-KPQVKw7vQnz_GAEXOLZ:bl13tjy) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/storage/raft_transport.go#L232)):*
> ```Go
> 		return t.processQueue(nodeID, ch, conn)
> 	}, 0); err != nil {
> 		if consecFailures == 0 {
> ```

When the breaker trips, we may want to drain all messages out of the queue and discard them (we may have accumulated several while trying to connect). Better to start fresh with new messages than replay (some) old ones. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-:-KPQVb6N1737Ql31-hns:bnfp4nl)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
@tamird the queue changes are specifically geared towards clarifying logging, which is what this PR is about. For queues, we were previously spamming logs with the message that we were adding a particular replica to the GC queue when it was already added. The change to queuing provides feedback as to whether the replica was newly added or already present.

---

Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[rpc/context.go, line 192 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQSQnRjL9V4GuBcy6I:-KPSd9pYlW8ub7m339ex:bha3ttw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/rpc/context.go#L192)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why does this lock need to be held here? all concurrent callers to GRPCDial will block while this is executing.
</details>
Because the `metaConn` structs in `ctx.conns.cache` are also accessed by the shutdown worker.

---

*[rpc/context.go, line 211 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQSgJK565OYPbv7yKO:-KPSdGMNn1Z76Yuh4IQa:biw9ugg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/rpc/context.go#L211)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this lock also doesn't appear necessary.
</details>
Ditto with above.

---

*[storage/client_raft_test.go, line 1175 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQSvq2PU9Mk4D9oAwN:-KPSeJbGyTubySUQzrRq:b-t4np4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/storage/client_raft_test.go#L1175)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> does this have anything to do with the rest of this change?
</details>
Just was a significant point of confusion when I was debugging logs from this test after it stopped working due to breaker changes. It's a drive by improvement.

---

*[storage/raft_transport.go, line 232 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQVKw6e5OocA43HrZS:-KPSeyWcqdbm2yS2fxhc:b-g21wby) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/storage/raft_transport.go#L232)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> When the breaker trips, we may want to drain all messages out of the queue and discard them (we may have accumulated several while trying to connect). Better to start fresh with new messages than replay (some) old ones. 
</details>
That already happens because this failure causes `connectAndProcess` to exit which causes the queue to be deleted when the worker finishes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-:-KPScUQ3OYGMtzikHI_j:b-41tbnf)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 18 of 23 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[GLOCKFILE, line 112 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQHlOjrxwuRo8UZvBp-r7-112:-KPSfLvoKnlfcFNEgx6Z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41c1f1d672dcfaf0527af39e6c0a889d4746c27c/GLOCKFILE#L112)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> This picks up the `DialContext` method.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 6 files at r9.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[rpc/context.go, line 192 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQSQnRjL9V4GuBcy6I:-KPQSR2fBr-N2kNiLqmY:b-5ia0u2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/rpc/context.go#L192)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Because the `metaConn` structs in `ctx.conns.cache` are also accessed by the shutdown worker.
</details>
What if the shutdown worker did something like
```
meta.Do(func() {})
```
before accessing `meta.conn`?

---

*[storage/client_test.go, line 273 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPSwR5d73hQTY_pjDgO:-KPSwR5d73hQTY_pjDgP:bohyc4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/9542f42a2a99e23c588056c68eb769132ae24728/storage/client_test.go#L273)):*
> ```Go
> 		// Create breaker options which retry very quickly and use a "real"
> 		// clock so breakers retry without requiring the manual clock to be
> 		// incremented manually to untrip breaker.
> ```

"the" breaker

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[rpc/context.go, line 192 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQSQnRjL9V4GuBcy6I:-KPTLmsbT-gBA-GEKzTG:b-gvxlym) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/rpc/context.go#L192)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> What if the shutdown worker did something like
> ```
> meta.Do(func() {})
> ```
> before accessing `meta.conn`?
</details>
I don't see how to avoid deadlocks or still have races.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529)*
<!-- Sent from Reviewable.io -->

tamird commented:
This has caused merge skew on master, which is no longer building.

---

Reviewed 1 of 6 files at r9.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[rpc/context.go, line 192 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-KPQSQnRjL9V4GuBcy6I:-KPTPez2iMKlQnYUwLBi:bvg3b5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/c49c995d0ac696c3580197d648716dc907a246e3/rpc/context.go#L192)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> I don't see how to avoid deadlocks or still have races.
</details>
```diff
 		ctx.cancel()
 		ctx.conns.Lock()
 		for key, meta := range ctx.conns.cache {
-			ctx.removeConnLocked(key, meta.conn)
+			ctx.removeConnLocked(key, meta)
 		}
 		ctx.conns.Unlock()
 	})
@@ -146,17 +146,18 @@ func (ctx *Context) SetLocalInternalServer(internalServer roachpb.InternalServer
 	ctx.localInternalServer = internalServer
 }
 
-func (ctx *Context) removeConn(key string, conn *grpc.ClientConn) {
+func (ctx *Context) removeConn(key string, meta *connMeta) {
 	ctx.conns.Lock()
 	defer ctx.conns.Unlock()
-	ctx.removeConnLocked(key, conn)
+	ctx.removeConnLocked(key, meta)
 }
 
-func (ctx *Context) removeConnLocked(key string, conn *grpc.ClientConn) {
+func (ctx *Context) removeConnLocked(key string, meta *connMeta) {
 	if log.V(1) {
 		log.Infof(ctx.masterCtx, "closing %s", key)
 	}
-	if conn != nil {
+	meta.Do(func() {}) // Make sure initialization is not in-progress.
+	if conn := meta.conn; conn != nil {
 		if err := conn.Close(); err != nil && !grpcutil.IsClosedConnection(err) {
 			if log.V(1) {
 				log.Errorf(ctx.masterCtx, "failed to close client connection: %s", err)
@@ -191,10 +192,8 @@ func (ctx *Context) GRPCDial(target string, opts ...grpc.DialOption) (*grpc.Clie
 			log.Infof(ctx.masterCtx, "dialing %s", target)
 		}
 		conn, err := grpc.DialContext(ctx.masterCtx, target, dialOpts...)
-		ctx.conns.Lock()
 		meta.conn = conn
 		meta.err = err
-		ctx.conns.Unlock()
 		if err == nil {
 			if ctx.Stopper.RunTask(func() {
 				ctx.Stopper.RunWorker(func() {
@@ -202,16 +201,14 @@ func (ctx *Context) GRPCDial(target string, opts ...grpc.DialOption) (*grpc.Clie
 					if err != nil && !grpcutil.IsClosedConnection(err) {
 						log.Error(ctx.masterCtx, err)
 					}
-					ctx.removeConn(target, conn)
+					ctx.removeConn(target, meta)
 				})
 			}) != nil {
-				ctx.removeConn(target, conn)
+				ctx.removeConn(target, meta)
 			}
 		}
 	})
 
-	ctx.conns.Lock()
-	defer ctx.conns.Unlock()
 	return meta.conn, meta.err
 }
``` 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-:-KPTPm9ug5ZRUKnI92WW:b-szlpqe)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Yikes. Forgot about that PR going in. Working on a fix now. Since most people aren't on master right now, I'm going to try to get the fix in instead of reverting.

tamird commented:
I have a PR out.

---

Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8529#-:-KPTRs7JEmflSQhqGy5J:bjaajpf)*
<!-- Sent from Reviewable.io -->
